### PR TITLE
Version 2: Fix certificate import dialog

### PR DESCRIPTION
**Pull request by OleksandrSPV**
https://github.com/PeculiarVentures/fortify-tools/assets/129727832/6f9882d4-9dcf-4766-8776-1a0c101cd577

### CODE REVIEW

File: src/utils/certificate.ts [hunk 1]
```diff
@@ -33,3 +38,52 @@ export function certificateSubjectToString(
...
+
+function base64Clarify(base64: string) {
+  const base64Re =
+    /-----BEGIN [^-]+-----([A-Za-z0-9+/=\s]+)-----END [^-]+-----|begin-base64[^\n]+\n([A-Za-z0-9+/=\s]+)====/;
+  const execArray = base64Re.exec(base64);
+
+  return execArray ? execArray[1] || execArray[2] : base64;
+}
+
+export const isHex = (value: string) =>
```
Comment (donskov) [inline src/utils/certificate.ts:64]: We have the `Convert.isHex` utility. Let's use it.

File: src/utils/certificate.ts [hunk 2]
```diff
@@ -33,3 +38,52 @@ export function certificateSubjectToString(
 
   return parts.join(", ");
 }
+
+export function bufferToString(buffer: ArrayBuffer) {
+  const bytes = new Uint8Array(buffer);
+  return bytes.reduce((string, byte) => string + String.fromCharCode(byte), "");
+}
+
+export const isBase64 = (value: string) => {
```
Comment (donskov) [inline src/utils/certificate.ts:47]: We have the `Convert.isBase64` utility. Let's use it.

**Review by donskov — CHANGES_REQUESTED** [review]

File: src/utils/certificate.ts [hunk 3]
```diff
@@ -33,3 +38,52 @@ export function certificateSubjectToString(
...
+    /-----BEGIN [^-]+-----([A-Za-z0-9+/=\s]+)-----END [^-]+-----|begin-base64[^\n]+\n([A-Za-z0-9+/=\s]+)====/;
+  const execArray = base64Re.exec(base64);
+
+  return execArray ? execArray[1] || execArray[2] : base64;
+}
+
+export const isHex = (value: string) =>
+  /^\s*(?:[0-9A-Fa-f][0-9A-Fa-f]\s*)+$/.test(value);
+
+export function certificateConvertRaw(cert: string | ArrayBuffer): ArrayBuffer {
```
Comment (donskov) [inline src/utils/certificate.ts:67]: Maybe we can use something like this https://github.com/PeculiarVentures/x509/blob/3a8fd4628be15cf96dfcd2f84173e82155f898d5/src/pem_data.ts#L9 instead. It includes the same functionality.

**Review by donskov — CHANGES_REQUESTED** [review]

File: src/utils/certificate.ts (hunk 3 above)
Comment (OleksandrSPV) [reply src/utils/certificate.ts:67]: Thanks, looks good, I'll check

Comment (OleksandrSPV) [reply src/utils/certificate.ts:67]: @donskov I tried to use Convert.isHex & Convert.isBase64 and it doesn't work for me. I understand now why guys from Goodkey didn't use it. Do you have any ideas?

Comment (OleksandrSPV) [reply src/utils/certificate.ts:62]: @donskov I've improved the converter method a bit here https://github.com/PeculiarVentures/fortify-tools/pull/150/commits/99549579e2c208169466b9b06931f07d483f9a86, but the previous question is still relevant.

File: src/dialogs/certificate-import-dialog/useCertificateImportDialog.tsx [hunk 4]
```diff
@@ -43,35 +44,77 @@ export function useCertificateImportDialog(props: {
...
   };
 
-  const onDropAccepted = (fileContent: ArrayBuffer) => {
-    const x509Cert = new X509Certificate(fileContent);
+  const onDropAccepted = (
+    fileContent: ArrayBuffer,
+    extension: string,
+    fileType: string
+  ) => {
+    const buf = certificateConvertRaw(fileContent);
```
Comment (donskov) [inline src/dialogs/certificate-import-dialog/useCertificateImportDialog.tsx:52]: I'm trying to figure out what exactly we need this function for. The `X509Certificate` class has checks to transpile input to the correct value.

Just tried `console.log("drop", new X509Certificate(fileContent));` without any extra transformations and see the correct output. Any thoughts?

**Review by donskov — CHANGES_REQUESTED** [review]

File: src/dialogs/certificate-import-dialog/useCertificateImportDialog.tsx (hunk 4 above)
Comment (OleksandrSPV) [reply src/dialogs/certificate-import-dialog/useCertificateImportDialog.tsx]: @donskov fixed here https://github.com/PeculiarVentures/fortify-tools/pull/150/commits/23dbd4fce74773f0c5f26481e8e2ca9fb8475ffd

**Review by donskov — APPROVED** [review]

**Pull request merged**